src/null-safety/migrating.md

```diff
@@ -0,0 +1,397 @@
+---
+
+This page describes how and when to migrate your code to [null safety][].
+Here are the basic steps for migrating each package that you own:
+
+1. _Optional but recommended:_ Wait for the packages
+   that you depend on to migrate.
+2. Migrate your package's code,
+   preferably using the interactive migration tool (`dart migrate`).
+3. Test to make sure your changes work.
+4. If the package is already on pub.dev,
+   publish the null-safe version.
+
+{{ site.alert.info }}
+  **Migrating an app is the same as migrating a package.**
+  Before fully migrating an app,
+  consider waiting until null safety is in a stable release.
+  In the meantime, you can add migration hints to your app
+  and take the opportunity to refactor your app code.
+  **_[PENDING: is that good/helpful advice?]_**
+{{ site.alert.end }}
+
+[null safety]: /null-safety
+
+![Screenshot of migration tool](/null-safety/migrating-to-null-safety/migration-tool.png)
+_This interactive tool (`dart migrate`) can simplify migration to null safety._
+
+**_[PENDING: flutter-specific stuff?
+Only things kevmoo's run into:
+--[no-]sound-null-safety (probably applies to `dart` as well)
+IDE launch options
+flutter run, flutter test flag order might be slightly different.
+pub run --enable-experiment=non-nullable test
+aka dart test --enable-experiment=non-nullable]_**
+
+## Prepare to migrate
+
+We strongly recommend migrating code in order, 
+with the leaves of the dependency graph being migrated first.
+For example, if package C depends on package B, which depends on package A,
+then A should be migrated to null safety first, then B, then C.
+
+![Illustration of C/B/A sentence](/null-safety/migrating/null-safety-migration-order.png){:width="454px"}<br>
+
+Although you _can_ migrate before your dependencies support null safety,
+you might have to change your code when your dependencies migrate.
+For example, if you predict that a function will take a nullable parameter but
+the package migrates it to be non-nullable,
+then passing a nullable argument becomes a compile error.
+
+
+### Switch to a beta release
+
+Start the migration process by switching to the **latest beta release**
+of either the Dart SDK or the Flutter SDK:
+
+* If you use the Dart SDK, download a beta release
+  from the [Dart SDK archive][].
+* If you use the Flutter SDK,
+  switch to the beta channel:
+
+  ```terminal
+$ flutter channel beta
+$ flutter upgrade
+```
+
+[Dart SDK archive]: /tools/sdk/archive#beta-channel
+
+{{ site.alert.warn }}
+  When you return to working on production code,
+  **switch back to a stable release**
+  (for example, by running `flutter channel stable`).
+{{ site.alert.end }}
+
+
+### Wait for dependencies
+
+Get the migration state of your package's dependencies
+by using the following command:
+
+```terminal
+$ dart pub outdated --mode=null-safety
+```
+
+If the output says **all the packages fully support null safety**,
+then you can start migrating.
+
```

Comment:
I've often been asked why both direct and dev dependencies need to be ready. We may want to elaborate with something like:

It is important that all direct dependencies are ready, as that is required to ensure you can run your app with sound null safety.

It is also important that all dev dependencies are ready, as that is required to be able to run tests with sound null safety.

Reply:
I have a question about this one (also in my draft, but since it might take me a little while to publish it): 

**Aren't there dev dependencies that have nothing to do with testing? E.g. deps you might need for analyzing or building your app**